server/src/main/java/org/opensearch/common/geo/GeoUtils.java

```diff
@@ -444,112 +448,201 @@ public static GeoPoint parseGeoPoint(XContentParser parser, GeoPoint point, fina
+            }
+        }
+
+        if (data.get(LONGITUDE) == null) {
+            throw new OpenSearchParseException("field [{}] missing", LONGITUDE);
+        }
+        if (data.get(LATITUDE) == null) {
+            throw new OpenSearchParseException("field [{}] missing", LATITUDE);
+        }
+
+        return point.reset(data.get(LATITUDE), data.get(LONGITUDE));
+    }
+
+    private static GeoPoint parseGeoHashFields(
+        final XContentParser parser,
+        final GeoPoint point,
+        final GeoUtils.EffectivePoint effectivePoint
+    ) throws IOException {
+        if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {
+            throw new OpenSearchParseException(ERR_MSG_INVALID_TOKEN, parser.currentToken());
+        }
+
+        if (!GEOHASH.equals(parser.currentName())) {
+            throw new OpenSearchParseException(ERR_MSG_INVALID_FIELDS);
+        }
+
+        if (parser.nextToken() != XContentParser.Token.VALUE_STRING) {
+            throw new OpenSearchParseException("{} must be a string", GEOHASH);
+        }
+
+        return point.parseGeoHash(parser.text(), effectivePoint);
+    }
+
+    private static GeoPoint parseGeoJsonFields(final XContentParser parser, final GeoPoint point, final boolean ignoreZValue)
+        throws IOException {
+        boolean hasTypePoint = false;
+        boolean hasCoordinates = false;
+        for (int i = 0; i < 2; i++) {
+            if (i != 0) {
+                parser.nextToken();
+            }
+
+            if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {
+                if (!hasTypePoint) {
+                    throw new OpenSearchParseException("field [{}] missing", GEOJSON_TYPE);
+                }
+                if (!hasCoordinates) {
+                    throw new OpenSearchParseException("field [{}] missing", GEOJSON_COORDS);
                 }
             }
-            if (geohash != null) {
-                if (!Double.isNaN(lat) || !Double.isNaN(lon)) {
-                    throw new OpenSearchParseException("field must be either lat/lon or geohash");
-                } else {
-                    return point.parseGeoHash(geohash, effectivePoint);
+
+            if (GEOJSON_TYPE.equals(parser.currentName())) {
+                if (parser.nextToken() != XContentParser.Token.VALUE_STRING) {
+                    throw new OpenSearchParseException("{} must be a string", GEOJSON_TYPE);
                 }
-            } else if (numberFormatException != null) {
-                throw new OpenSearchParseException("[{}] and [{}] must be valid double values", numberFormatException, LATITUDE, LONGITUDE);
-            } else if (Double.isNaN(lat)) {
-                throw new OpenSearchParseException("field [{}] missing", LATITUDE);
-            } else if (Double.isNaN(lon)) {
-                throw new OpenSearchParseException("field [{}] missing", LONGITUDE);
+
+                if (!GEOJSON_TYPE_POINT.equals(parser.text())) {
```

Comment:
See my above note on case sensitivity. 

```suggestion
                if (ShapeType.forName(parser.text()) != ShapeType.POINT) {
```